nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,565 @@
+
+```javascript
+// app.js
+const express = require("express");
+const app = express();
+const usersRouter = require("./routes/usersRouter");
+const usersStorage = require("./storages/usersStorage");
+
+app.set("view engine", "ejs");
+app.use(express.urlencoded({ extended: true }));
+app.use("/users", usersRouter);
+app.get("/", (req, res) => res.render("index", { title: "User List", users: usersStorage.getUsers() }));
+
+const PORT = process.env.PORT || 3000;
+app.listen(PORT, () => console.log(`Express app listening on port ${PORT}!`));
+```
+
+Most simple forms will use the `application/x-www-form-urlencoded` `Content-Type` when sending data to the server. Express, however, can't natively parse that data. We can use the `express.urlencoded()` middleware to handle this for us. When `extended` is `false`,  our server will only accept a `string` or an array of data, so we set it to `true` for some added flexibility. Note that if the `Content-Type` doesn't match `application/x-www-form-urlencoded`, then your server will show the data as an empty object `{}`.
+
+Let's create a new router called `usersRouter.js` in the routes folder:
+
+```javascript
+// routes/usersRouter.js
+const { Router } = require("express");
+const usersController = require("../controllers/usersController");
+const usersRouter = Router();
+
+usersRouter.get("/create", usersController.usersCreateGet);
+usersRouter.post("/create", usersController.usersCreatePost);
+
+module.exports = usersRouter;
+```
+
+Next we'll create two views:
+
+- `index.ejs` will list all the users we've created.
+- `users.ejs` will display our main form.
+
+```ejs
+<!-- views/index.ejs -->
+<!DOCTYPE html>
+<html>
+<body>
+<h1><%= title %></h1>
+<ul>
+  <% if (locals.users) {%>
+    <% users.forEach(user => { %>
+      <li>
+        ID: <%= user.id %>, Name: <%= user.firstName %> <%= user.lastName %>
+      </li>
+    <% }); %>
+  <% } %>
+</ul>
+<a href="/users/create">Create a user</a>
+</body>
+</html>
+```
+
+```ejs
+<!-- views/users.ejs -->
+<!DOCTYPE html>
+<html>
+<head>
+  <title><%= title %></title>
+</head>
+<body>
+<h1><%= title %></h1>
+<form action="/users/create" method="POST">
+  <label for="firstName">First Name: </label>
+  <input type="text" name="firstName" required>
+  <label for="lastName">Last Name: </label>
+  <input type="text" name="lastName" required>
+  <button type="submit">Submit</button>
+</form>
+<a href="/">Back to home</a>
+</body>
+</html>
+```
+
+The logic for this router will go inside of our controller:
+
+```javascript
+// controllers/usersController.js
+// User is a factory function which acts as a model for our simulated database.
+// We will explore this concept more as we later dive deeper into databases.
+const User = (firstName, lastName) => ({ firstName, lastName });
+const asyncHandler = require("express-async-handler");
+const usersStorage = require("../storages/usersStorage");
```

Comment:
Since you've refactored to use a `usersStorage` class, I reckon the `User` factory can be an internal implementation detail for that class. No need to have it be external. What do you think?

Reply:
I did give this some thought and wasn't sure how relevant I wanted the User factory to be, since it's not really the focal point. But given that the `usersStorage` class was split off, I was considering either a `models/user.js` or like you said just embedding it within the class directly. How do you feel about either approach?

Reply:
Learners haven't been introduced to the "models" as a concept yet, so I wouldn't want `modles/user.js`. Just making it internal to the `userStorage` class should be good enough.